File: java/v2styx-lib/src/main/java/com/v2soft/styxlib/l5/serialization/impl/BufferWriterImpl.java
package com.v2soft.styxlib.l5.serialization.impl;

import com.v2soft.styxlib.exceptions.StyxException;
import com.v2soft.styxlib.l5.serialization.IBufferWriter;
import com.v2soft.styxlib.l5.dev.MetricsAndStats;

import java.nio.BufferOverflowException;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;

public class BufferWriterImpl implements IBufferWriter {
    private static final int sDataBufferSize = 16;
    protected static final Charset sUTFCharset = StandardCharsets.UTF_8;
    private final byte[] mInternalBuffer;
    private final ByteBuffer mBuffer;

    public BufferWriterImpl(int size) {
        mInternalBuffer = new byte[sDataBufferSize];
        MetricsAndStats.byteArrayAllocationIo++;
        mBuffer = ByteBuffer.allocate(size);
        MetricsAndStats.byteBufferAllocation++;
    }

    @Override
    public void writeUInt8(short val) throws StyxException {
        mInternalBuffer[0] = (byte) val;
        write(mInternalBuffer, 0, 1);
    }

    @Override
    public void writeUInt16(int val) throws StyxException {
        mInternalBuffer[0] = (byte) ((byte) val & 0xFF);
        mInternalBuffer[1] = (byte) ((byte) (val >> 8) & 0xFF);
        write(mInternalBuffer, 0, 2);
    }

    @Override
    public void writeUInt32(long val) throws StyxException {
        mInternalBuffer[0] = (byte) ((byte) val & 0xFF);
        mInternalBuffer[1] = (byte) ((byte) (val >> 8) & 0xFF);
        mInternalBuffer[2] = (byte) ((byte) (val >> 16) & 0xFF);
        mInternalBuffer[3] = (byte) ((byte) (val >> 24) & 0xFF);
        write(mInternalBuffer, 0, 4);
    }

    @Override
    public void writeUInt64(long val) throws StyxException {
        mInternalBuffer[0] = (byte) ((byte) val & 0xFF);
        mInternalBuffer[1] = (byte) ((byte) (val >> 8) & 0xFF);
        mInternalBuffer[2] = (byte) ((byte) (val >> 16) & 0xFF);
        mInternalBuffer[3] = (byte) ((byte) (val >> 24) & 0xFF);
        mInternalBuffer[4] = (byte) ((byte) (val >> 32) & 0xFF);
        mInternalBuffer[5] = (byte) ((byte) (val >> 40) & 0xFF);
        mInternalBuffer[6] = (byte) ((byte) (val >> 48) & 0xFF);
        mInternalBuffer[7] = (byte) ((byte) (val >> 56) & 0xFF);
        write(mInternalBuffer, 0, 8);
    }

    @Override
    public void writeUTFString(String string) throws StyxException {
        byte[] data = string.getBytes(sUTFCharset);
        writeUInt16(data.length);
        write(data, 0, data.length);
    }

    @Override
    public int write(byte[] data, int offset, int count) throws StyxException {
        try {
            mBuffer.put(data, offset, count);
        } catch (BufferOverflowException err) {
            String message = String.format(
                    "Buffer overflow: Buffer: c=%d, p=%d, l=%d, Write: c=%d, o=%d, l=%d",
                    mBuffer.capacity(), mBuffer.position(), mBuffer.limit(),
                    count, offset, data.length);
            throw new com.v2soft.styxlib.l5.serialization.BufferOverflowException(message);
        }
        return count;
    }

    @Override
    public void prepareBuffer(int bufferSize) {
        mBuffer.clear();
        mBuffer.limit(bufferSize);
    }

    @Override
    public ByteBuffer getBuffer() {
        return mBuffer;
    }

    @Override
    public int getPosition() {
        return mBuffer.position();
    }

    @Override
    public int getLimit() {
        return mBuffer.limit();
    }
}

File: cpp/modules/l5/src/BufferWriterImpl.cpp
#include "serialization/BufferWriterImpl.h"
#include <stdexcept>

BufferWriterImpl::BufferWriterImpl(size_t bufferSize)
{
    buffer = new uint8_t[bufferSize];
    position = 0;
    limit = bufferSize;
}

BufferWriterImpl::~BufferWriterImpl()
{
    delete[] buffer;
}

void BufferWriterImpl::writeUInt8(uint8_t val)
{
    if (position + sizeof(uint8_t) > limit)
    {
        throw std::out_of_range("Buffer overflow");
    }
    buffer[position++] = val;
}
void BufferWriterImpl::writeUInt16(uint16_t val)
{
    if (position + sizeof(uint16_t) > limit)
    {
        throw std::out_of_range("Buffer overflow");
    }
    buffer[position++] = static_cast<uint8_t>(val & 0xFF);
    buffer[position++] = static_cast<uint8_t>(val >> 8);
}
void BufferWriterImpl::writeUInt32(uint32_t val)
{
    if (position + sizeof(uint32_t) > limit)
    {
        throw std::out_of_range("Buffer overflow");
    }
    buffer[position++] = static_cast<uint8_t>(val & 0xFF);
    buffer[position++] = static_cast<uint8_t>(val >> 8);
    buffer[position++] = static_cast<uint8_t>(val >> 16);
    buffer[position++] = static_cast<uint8_t>(val >> 24);
}
void BufferWriterImpl::writeUInt64(uint64_t value)
{
    if (position + sizeof(uint64_t) > limit)
    {
        throw std::out_of_range("Buffer overflow");
    }
    buffer[position++] = static_cast<uint8_t>(value & 0xFF);
    buffer[position++] = static_cast<uint8_t>(value >> 8);
    buffer[position++] = static_cast<uint8_t>(value >> 16);
    buffer[position++] = static_cast<uint8_t>(value >> 24);
    buffer[position++] = static_cast<uint8_t>(value >> 32);
    buffer[position++] = static_cast<uint8_t>(value >> 40);
    buffer[position++] = static_cast<uint8_t>(value >> 48);
    buffer[position++] = static_cast<uint8_t>(value >> 56);
}
void BufferWriterImpl::writeUTFString(StyxString string)
{
    size_t len = string.length();
    if (len > limit - position - sizeof(uint16_t))
    {
        throw std::out_of_range("Buffer overflow");
    }
    writeUInt16(static_cast<uint16_t>(len));
    for (size_t i = 0; i < len; ++i)
    {
        buffer[position++] = static_cast<uint8_t>(string[i]);
    }
}

StyxSize BufferWriterImpl::write(const StyxBuffer data, StyxSize count)
{
    if (position + count > limit)
    {
        throw std::out_of_range("Buffer overflow");
    }
    for (StyxSize i = 0; i < count; ++i)
    {
        buffer[position++] = data[i];
    }
    return count;
}
void BufferWriterImpl::prepareBuffer(StyxSize bufferSize)
{
    delete[] buffer;
    buffer = new uint8_t[bufferSize];
    position = 0;
    limit = bufferSize;
}

StyxBuffer BufferWriterImpl::getBuffer() const
{
    return buffer;
}